### ISSUE TITLE
Cleanup database functions and come up with some generic ones

### ISSUE THREAD

**Issue by stgraber** [lxc/lxd#2352]
While working on #2349 I've ended up having to do a whole bunch of copy/paste from other database functions. The amount of code duplication around there may become a source of issues for us in the future so we may want to look at cleaning this up a bit.

Specifically:
 - Keep db* functions limited to DB handling. Input validation, duplicate entries detection, ... is all supposed to happen before we hit them. Those functions are just a shim between Go and SQL.
 - Make sure we're consistent in how we handle errors and queries that return no entries.
 - Come up with some generic shared functions that we can use for the most common cases (get all the entry names from table X, delete an entry, get an entry's ID from its name, get an entry's name from its ID, ...)

That should make things a bit more maintainable going forward.

**Comment by techtonik** [comment]
Isn't there any ready to use ORM for go? https://github.com/avelino/awesome-go#database SQL query builder section.

**Comment by stgraber** [comment]
Last we checked, nothing that handles sqlite in a sane way. We ran into a whole lot of complex threading issues with sqlite that we fixed upstream by adding a bunch of custom flags which basically nobody else supports.

**Comment by stgraber** [comment]
@freeekanayaka Giving that one to you as you've pretty much been doing this work already :)

**Comment by freeekanayaka** [comment]
@stgraber sure thing :) I think we should indeed try to avoid ORMs (even regardless of how much sqlite-aware are), at the end of the day they usually don't buy that much and reduce your freedom. What you generally want is to come up with your own "ORM" (as we effectively have, despite it's not very DRY), which is tailored for the app.

**Comment by stgraber** [comment]
Yup, absolutely. What we want are convenient helpers that match our DB handling patterns.

The LXD database is pretty simple and so are our requests. We also try to be pretty consistent so there's a lot of similarities between the various DB tables. That's where those helpers come into play and make it easy to read/write data without re-inventing the wheel with every new object collection we have to store in the DB.

Looks like you're well underway to doing this already as part of the dqlite work, so we should be able to call this issue done soon.

**Comment by techtonik** [comment]
I don't mind against DRY databases as long as they don't require C compiler like `sqlite`.
